pandas/core/groupby/generic.py

```diff
@@ -1754,10 +1757,27 @@ def value_counts(
                     level=index_level, sort_remaining=False
                 )
 
-            if not self.as_index:
+            if self.as_index:
+                return result.__finalize__(self.obj, method="value_counts")
+            else:
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
-            return result.__finalize__(self.obj, method="value_counts")
+                name = "proportion" if normalize else "count"
```

Comment:
i dont' understand why you are reinventing all kinds of code here. can you not just concat?

Reply:
I tried! AFAIK concat does not do what is required here. I eplained why not in my comments above:

> I could not use `Series.to_frame()` because:
> 
> * It throws away duplicate level names
> * MultiIndex has thrown away some dtypes, so I need to set each column
> 
> I could have used `concat`, but then I would have had to add the name and the index and then remove the index at the end. `insert` was simpler and neater.